server/src-lib/Hasura/RQL/DDL/RemoteRelationship/Validate.hs

```diff
@@ -409,7 +412,16 @@ isTypeCoercible actualType expectedType =
   let (actualBaseType, actualNestingLevel) = getBaseTyWithNestedLevelsCount actualType
       (expectedBaseType, expectedNestingLevel) = getBaseTyWithNestedLevelsCount expectedType
   in
-  if | actualBaseType /= expectedBaseType -> raiseValidationError
+  if | expectedBaseType == $$(G.litName "ID") ->
+         bool (throwError $ IDTypeJoin actualBaseType)
+              (pure ())
+              -- Check under `Input Coercion` https://spec.graphql.org/June2018/#sec-ID
+              -- We can also include the `ID` type in the below list but it will be
+              -- extraneous because at the time of writing this, we don't generate
+              -- the `ID` type in the DB schema
+              (G.unName actualBaseType `elem`
+               [ "Int", "String", "bigint", "smallint" ])
```

Comment:
```suggestion
          -- Check under `Input Coercion` https://spec.graphql.org/June2018/#sec-ID
          -- We can also include the `ID` type in the below list but it will be
          -- extraneous because at the time of writing this, we don't generate
          -- the `ID` type in the DB schema
         unless (actualBaseType `elem` [ "Int", "String", "bigint", "smallint" ]) $
           throwError $ IDTypeJoin actualBaseType
```
since `Name` is instance `IsString` here you can omit the `unName` (I think!)